File: Data Structures & Algorithms/longest-increasing-subsequence/submission-6.java
class Solution {
    Integer[][] cache;
    public int lengthOfLIS(int[] nums) {
        cache = new Integer[nums.length][nums.length+1];
        return dfs(nums, 0, -1);
    }

    private int dfs(int[] nums, int cur, int prev) {

        if(cur >= nums.length)
            return 0;
        if(cache[cur][prev+1] != null) return cache[cur][prev+1];
        int include = 0;
        if(prev == -1 || nums[cur] > nums[prev]) 
            include = 1 + dfs(nums, cur+1, cur);
        int skip = dfs(nums, cur+1, prev);

        cache[cur][prev+1] = Math.max(include, skip);
        return cache[cur][prev+1];

    }
}

File: Data Structures & Algorithms/longest-consecutive-sequence/submission-6.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    longestConsecutive(nums) {
        const s = new Set();
        if(nums.length == 0) return 0;
        for(let n of nums) {
            s.add(n);
        }

        let min = Math.min(...nums); 
        let temp = min;
        let count = 1;
        while(s.has(temp+1)) {
            count++;
            temp++;
        }
        let maxCount = count;
        s.delete(min);
        count = 1;
        for(let cur of s) {
            while(s.has(cur+1)) {
                cur++;
                count++;
            }
            maxCount = Math.max(maxCount, count);
            count = 1;

        }
        maxCount = Math.max(maxCount, count);
        return maxCount;

    }
}
